arm9/source/utils/scripting.c

```diff
@@ -1710,56 +1813,181 @@ bool MemTextViewer(const char* text, u32 len, u32 start, bool as_script) {
+                *((char *) cursor++) = key_character;
+                if (key_character == '\r') {
+                    memmove((char *) cursor + 1, cursor, text + len++ - cursor + 1);
+                    *((char *) cursor++) = '\n';
+                }
+            }
+
+            if (cursor && !ww) {
+                const char* cursor_line_start = line_start(text, len, ww, cursor);
+                u32 cursor_chars_from_line_start = chars_between_pointers(cursor_line_start, cursor);
+                if (cursor_chars_from_line_start < off_disp_chars + strlen(al_str)) off_disp_chars = cursor_chars_from_line_start - strlen(al_str);
+                if (cursor_chars_from_line_start >= off_disp_chars + TV_LLEN_DISP - strlen(ar_str)) off_disp_chars = cursor_chars_from_line_start + strlen(ar_str) - TV_LLEN_DISP + 1;
+            }
+            while (cursor && cursor < line0_next) line0_next = line_seek_chars(text, len, ww, line0_next, -1);
+            while (cursor && line0_next < line_seek_chars(text, len, ww, GetNextChar(cursor), -TV_NLIN_DISP)) line0_next = line_seek_chars(text, len, ww, line0_next, 1);
+        }
+
+        // find last allowed lines (ww and nonww)
+        const char* llast_nww = line_seek_chars(text, len, 0, text + len + 1, -TV_NLIN_DISP);
+        const char* llast_ww = line_seek_chars(text, len, TV_LLEN_DISP, text + len + 1, -TV_NLIN_DISP);
 
         // check for problems, apply changes
         if (!ww && (line0_next > llast_nww)) line0_next = llast_nww;
         else if (ww && (line0_next > llast_ww)) line0_next = llast_ww;
         if (line0_next < line0) { // fix line number for decrease
-            do if (*(--line0) == '\n') lcurr--;
+            do {
+                DecChar(&line0);
+                if (is_newline(line0)) lcurr--;
+            }
             while (line0 > line0_next);
         } else { // fix line number for increase / same
-            for (; line0_next > line0; line0++)
-                if (*line0 == '\n') lcurr++;
+            for (; line0_next > line0; IncChar(&line0)) if (is_newline(line0)) lcurr++;
+        }
+
+        // find maximum line length
+        u32 llen_max = 0;
+        for (const char* ptr = text; ptr < text + len; ptr = line_seek_chars(text, len, 0, ptr, 1)) {
+            u32 llen = line_len_chars(text, len, 0, ptr, NULL) + 1;
+            if (llen > llen_max) llen_max = llen;
         }
-        if (off_disp + TV_LLEN_DISP > llen_max) off_disp = llen_max - TV_LLEN_DISP;
-        if ((off_disp < 0) || ww) off_disp = 0;
+
+        if (off_disp_chars + TV_LLEN_DISP > llen_max) off_disp_chars = llen_max - TV_LLEN_DISP;
+        if (off_disp_chars < 0 || ww) off_disp_chars = 0;
+    }
+
+    // check for user edits
+    if (text_cpy) {
+        if (save_path) {
+            bool diffs = false;
+            if (len != text_cpy_len) diffs = true;
+            else for (u32 i = 0; i < len; ++i) if (text[i] != text_cpy[i]) { diffs = true; break; }
```

Comment:
This could be slow for large files, but I took it form the hex editor; open to suggestions here...